8382 style per node v2

javitonino opened the pull request:
This persists layer style per analysis node (in a new table), so when deleting an analysis, it goes back to the old style. Summary:
- On each layer update, the backend saves the layer style to its corresponding entry on the LayerNodeStyle table. It is indexed by layer_id and analysis nodes, because multiple layers can point to the same node with different styles.
- This information is added to the analysis presenters to share it with the front.
- When doing an operation that needs to recover styles (e.g: deleting the latest analysis), the frontend restores the styles from the analysis history, and PUTs then to the layer controller (along with the source_id change).

This works fine for deleting analysis nodes. The style restoration is done in the frontend so we know what to paint without waiting for the backend response, otherwise we would have to freeze the interface.

There is one other case that complicates this. When a node is dragged out of a layer, we need to copy the style history from the old layer to the new one. This is necessary so the deletion of analysis in the newly created layer correctly restores styles.

This is done in backend, in LayersController.create, using a new parameter send from frontend (from_layer_id). It is a bit complicated because the copy needs to rename the nodes to the new layer letter, and we have to rename a node in the old layer as well.

~~There are still some problems with some of the layer deletion flows where there are dependencies between layers, and analysis nodes gets moved from layer to layer. It is probably not worth to fix those with this approach, as it only happens in an uncommon case and it will mess up the code even more (we would have to use an approach similar to what's described in the previous paragraph). We have decided not to this until we can move user actions to backend.~~

Also, in the analysis controller we have to invalidate the cached styles each time a param of an analysis changes, as that could change the schema and invalidate the styles. As we don't have an smarter approach, we unconditionally invalidate, so each change to a analysis nodes purges the styles for that node and their parents.

In fact, the style keeping should probably be done entirely on frontend or backend. The problem right now is that it should be frontend (is where this operations are performed), and frontend people are busy + we will eventually :tm: move this node manipulation to the backend.

Fixes #8382

javitonino commented:
Hi, @juanignaciosl. Can you take a look at the explanation (and maybe at the code) and give your opinion. I'm not asking for a CR on the code itself (it needs to be cleaned up yet), but the general approach.

juanignaciosl commented:
Well, it's quite complex because of the constraints (keeping operations in front to avoid freezing, not doing "backend actions" yet) and so on, but the approach should work. Great explanation, including whys, why not and so on 👌 

I guess that "layer update" also includes some node operations, such as adding one, isn't it?

javitonino commented:
I do not touch analysis nodes at all in this PR, only the new LayerNodeStyle model. All node operations are still done by frontend as always. When updating a layer, I update/create the corresponding LayerNodeStyle model, but I trust that the actual node is alright without verification.

Forgot to mention, the reason I use a new model, and I do not have any source_id validation is #10221. When you are deleting the latest analysis node from a layer the process is:
 - DELETE /analysis/:id
 - PUT /layers/:id
 - POST /analyses

juanignaciosl commented:
That deletion order makes impossible storing the styles either in `layers` or in `analyses`?

javitonino commented:
I cannot use analyses (initial approach), but I could store it in the layer. The reason it's not in options is to avoid having the frontend sending/receiving that on every requests. Discarding that, I went for a separate model rather than creating a new JSON field for this.

I like it better as it makes source_id a top-level columnd, which should make validation easier when we can start doing it (when we start moving some user actions to back).

javitonino commented:
I think this is ready to see some testing. Changes since the previous RFC:
 - The styles are sent in the analysis presenter instead of the layer presenter.
 - Each change to an analysis node, invalidates the style *cache* for that node and all that depend on it

The first change is made to support the second one. The second is needed because changing analysis parameters can alter the returned schema and invalidate the styles. e.g: you create a centroid analysis with categories and style based on the categories. Later, you update that analysis to avoid using categories, so the style needs to be invalidated.

CR @xavijam @juanignaciosl 

This is deployed to staging for testing.

juanignaciosl commented on app/controllers/api/json/layers_controller.rb:
```diff
@@ -56,6 +56,32 @@ def create
+                node_id_to_fix = @layer.options['source'].gsub(letter, from_letter)
+                style_node = ::LayerNodeStyle.where(layer_id: from_layer.id, source_id: node_id_to_fix).first
+                if style_node
+                  style_node.source_id = @layer.options['source']
```
Should "old" ("from") layer `source_id` _always_ change if a new layer is created from an existing node?

juanignaciosl commented on app/controllers/carto/api/analyses_controller.rb:
```diff
@@ -55,6 +62,32 @@ def destroy
       private
 
+      def purge_layer_node_style_cache(node_ids)
+        Carto::LayerNodeStyle.from_visualization_and_source(@visualization, node_ids).delete_all
```
Ok, `from_visualization_and_source` accepts both single and multiple (array) `node_ids` thanks to ActiveRecord _magic_ 🙊

javitonino replied:
This is one of these things that I discovered it worked by accident. It works and doesn't feel too bad, so I decided to keep it. But just say the word and I'll change it to `from_visualization_and_sources` and pass `[node_id]` for the case where looking for a single entry.

juanignaciosl commented on app/controllers/carto/api/analyses_controller.rb:
```diff
@@ -43,8 +43,15 @@ def create
+
+        @analysis.analysis_definition = new_definition
         @analysis.save!
+        purge_layer_node_style_cache(affected_node_ids)
```
Isn't _cache_ a little misleading? Isn't this a _history_?

juanignaciosl commented on app/models/layer.rb:
```diff
@@ -85,6 +86,7 @@ def after_save
     maps.each(&:invalidate_vizjson_varnish_cache)
 
     register_table_dependencies if data_layer?
+    update_layer_node_style if DATA_LAYER_KINDS.include?(kind)
```
IMHO `dala_layer?` should be used instead of `DATA_LAYER_KINDS`.

javitonino replied:
Sure! This predates that change (in fact, I opened that ticket while writing this line) and forgot to change it. Thanks for the catch!

juanignaciosl commented on app/models/layer.rb:
```diff
@@ -226,4 +228,21 @@ def map
+  def update_layer_node_style
+    style = current_layer_node_style
+    if style
+      style.update_from_layer
```
I doubt whether passing arguments (maybe the full `layer`) would be better. There's something weird about asking the style to update itself but then needing to read the layer. Minor, anyway.

juanignaciosl commented on app/models/layer_node_style.rb:
```diff
@@ -0,0 +1,21 @@
+  OPTIONS_TO_COPY = [:sql_wrap, :style_properties, :tile_style].freeze
+
+  def duplicate
+    lns = LayerNodeStyle.new(source_id: source_id)
```
Can't you pass all 4 attributes at the constructor?

javitonino replied:
Let me double check this. I think I did this because of some serialization issues, but it may not be needed anymore.

javitonino replied:
Removed it altogether with the latest changes, it is not needed anymore.

javitonino commented:
This is merged with branch `issue-10409`

javitonino commented:
@xavijam @juanignaciosl I think this is as ready as it's going to be. I think it's time for CR.

juanignaciosl commented on app/models/layer.rb:
```diff
@@ -226,4 +230,21 @@ def map
+
+  def current_layer_node_style
+    return nil unless source_id
+    LayerNodeStyle.find(layer_id: id, source_id: source_id) || LayerNodeStyle.new(layer: self, source_id: source_id)
```
How come this is not based in `(self.)layer_node_styles`?

javitonino replied:
I think there was a good reason regarding Sequel caching, but let me reevaluate, as I changed a lot of code around here.

juanignaciosl commented on db/migrate/20161025084350_create_layer_node_style.rb:
```diff
@@ -0,0 +1,20 @@
+      json        :options, null: false
+      json        :infowindow, null: false
+      json        :tooltip, null: false
+      unique      [:layer_id, :source_id]
```
Nice.

juanignaciosl commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "cartodb-ui",
-  "version": "4.3.01",
+  "version": "4.4.02j",
```
👀

javitonino replied:
Almost impossible to forget, as GH complains about merging conflicts :sweat_smile:

javitonino commented on app/controllers/carto/api/analyses_controller.rb:
```diff
@@ -55,6 +62,32 @@ def destroy
+        kept_ids.select do |node_id|
+          old_node = old_nodes.find { |n| n.id == node_id }
+          new_node = new_nodes.find { |n| n.id == node_id }
+          old_node.non_child_params != new_node.non_child_params
```
Yes in practice, but I shouldn't. I'll add a sort around here.

javitonino commented on app/controllers/api/json/layers_controller.rb:
```diff
@@ -164,4 +168,48 @@ def validate_read_write_permission
+          end
+        end
+      end
+      update_source_layer_styles(from_layer, from_letter, to_letter, to_source)
```
Probably much nicer. This just kind of organically grow as I added more logic here. A bit of order would be nice :+1:

javitonino commented on app/controllers/api/json/layers_controller.rb:
```diff
@@ -164,4 +168,48 @@ def validate_read_write_permission
+        if lns.source_id.starts_with?(from_letter)
+          if lns.source_id[1..-1].to_i < to_source[1..-1].to_i
+            # Move LayerNodeStyles from the old layer if given.
+            lns.source_id = new_id
```
If you look at the following line, it adds the LayerNodeStyle to the new layer. This is just equivalent to `lns.layer_id = to_layer.id && lns.save` (plus fixes the Sequel cache). So what this does is move the node from the old layer to the new one, changing the letter in the name from the one in the source layer to the one in the to layer

javitonino commented on app/controllers/api/json/layers_controller.rb:
```diff
@@ -164,4 +168,48 @@ def validate_read_write_permission
     render_jsonp({ description: e.message }, 400)
   end
+
+  def copy_layer_node_styles(to_layer, from_layer, from_letter)
```
I prefer to keep it in the controller as it is quite specific for a single case, and I think it makes more sense not to pollute the model with something that only is used for a single operation, and it's a bit hacky (to say the least).

But no strong feelings either way, I could move it.
